src/sentry/utils/snowflake.py

```diff
@@ -55,3 +94,55 @@ def snowflake_id_generation():
     ID_VALIDATOR.validate(snowflake_id)
 
     return snowflake_id
+
+
+def get_redis_cluster(redis_key: str):
+    return redis.clusters.get("default").get_local_client_for_key(redis_key)
+
+
+def store_snowflake_to_redis(redis_key: str, timestamp: int, used_region_sequence: int) -> None:
+    cluster = get_redis_cluster(redis_key)
+    used_region_sequences = cluster.get(str(timestamp))
+    if used_region_sequences:
+        used_region_sequences = json.loads(used_region_sequences)
```

Comment:
Do you need JSON here? Would retaining only the 'previous' value allow you to increment and move on? You should be able to use the [incr](https://redis.io/commands/incr/) operation in redis to bump a counter atomically and get the new value. INCR will handle new time segments automatically as it defaults keys to 0. We would still need to constrain the upper value to prevent overflows though.

Reply:
hmm since it's atomic, this will erase a lot of the race conditions. If we use incr instead the new flow for region sequence will look like this:

1. check Redis to see if there is a value for the current timestamp, if not start at 0, if there is incr by 1
2. if the current value is 15, we want to look at the past and repeat the steps at the new timestamp

As a result of bumping the counter atomically, we don't have to worry about generating a random order since by nature multiple servers won't select the same sequence value